pwchem/protocols/VirtualDrugScreening/protocol_ligand_characterization.py

```diff
@@ -0,0 +1,132 @@
+from pwchem.objects import SetOfSmallMolecules, SmallMoleculesLibrary
+from pwchem.constants import DESCRIPTOR_CATEGORIES
+
+class ProtocolLigandCharacterization(EMProtocol):
+    _label = 'Ligand characterization'
+
+    def _defineParams(self, form):
+        form.addSection(label='Parameters')
+        form.addParam('inputSmallMolecules', params.PointerParam,
+                      pointerClass='SetOfSmallMolecules, SmallMoleculesLibrary', allowsNull=False,
+                      label="Input Small Molecules",
+                      help='Set of molecules to process.')
+        # Name all parameters using booleans
+        form.addParam('useConstitutional', params.BooleanParam, default=True,
+                    label='Calculate constitutional descriptors')
+        form.addParam('useElectronic', params.BooleanParam, default=True,
+                    label='Calculate electronic descriptors')
+        form.addParam('useTopological', params.BooleanParam, default=True,
+                    label='Calculate topological descriptors')
+        form.addParam('useGeometrical', params.BooleanParam, default=True,
+                    label='Calculate geometrical descriptors')
+        form.addParam('useRing', params.BooleanParam, default=True,
+                    label='Calculate ring descriptors')
+        form.addParam('useFragment', params.BooleanParam, default=True,
+                    label='Calculate fragment descriptors')
+        form.addParam('useOther', params.BooleanParam, default=True,
+                    label='Calculate other descriptors')
+
+    def _insertAllSteps(self):
+        self._insertFunctionStep('runDescriptorCalc')
+        self._insertFunctionStep('createOutputStep')
+
+    def runDescriptorCalc(self):
+        mol_set = self.inputSmallMolecules.get()
+
+        # Build molecule dictionary: {molName: path}
+        mol_dict = {mol.molName.get(): mol.getFileName() for mol in mol_set}
+        input_json = self._getExtraPath("input_mols.json")
+        with open(input_json, 'w') as f:
+            json.dump(mol_dict, f)
+
+        # Build descriptor flags
+        flags = {cat: getattr(self, f"use{cat.capitalize()}") for cat in DESCRIPTOR_CATEGORIES}
+        flags_json = json.dumps(flags)
+
+        # Save flags as file (optional, not used directly by script)
+        flags_path = self._getExtraPath("descriptor_flags.json")
+        with open(flags_path, 'w') as f:
+            f.write(flags_json)
+
+        # Output file path
+        output_json = self._getExtraPath("output_results.json")
+
+        # Call the RDKit script in the correct environment
+        Plugin.runScript(self, 'ligand_descriptor_calc.py',
+                        f"{input_json} {output_json} '{flags_json}'",
+                        env=RDKIT_DIC,
+                        cwd=self._getPath())
+
+        # Load the results
+        with open(output_json, 'r') as f:
+            data = json.load(f)
+
+        self.header = data['header']
```

Comment:
Almost there, the protocol seems fine, but we try not to store protocol attributes that will be used across steps. That means, no defining self.whatever in a step and using it in the next. In this case, the results of the script are properly saved in a file, the parsing must be done directly in the output step. Otherwise, the protocol feels correct